mangaki/mangaki/utils/recommendations.py

```diff
@@ -49,27 +51,24 @@ def get_reco_algo(request, algo_name='knn', category='all'):
     chrono.save('get rated works')
 
     if algo_name == 'knn':
-        queryset = Rating.objects.filter(work__in=already_rated_works)
-        dataset = Dataset()
-        triplets = list(
-            queryset.values_list('user_id', 'work_id', 'choice'))
-        if request.user.is_anonymous:
-            triplets.extend([
-                (request.user.id, work_id, choice)
-                for work_id, choice in current_user_ratings(request).items()
-            ])
-
-        anonymized = dataset.make_anonymous_data(triplets)
+        try:
+            algo = get_algo_backup(algo_name)
+            dataset = get_dataset_backup(algo_name)
+        except FileNotFoundError:
+            triplets = list(
+                Rating.objects.values_list('user_id', 'work_id', 'choice'))
+            chrono.save('get all %d interesting ratings' % len(triplets))
+            dataset, algo = fit_algo(algo_name, triplets)
+        already_rated_works['rating'] = already_rated_works['choice'].map(rating_values)
```

Comment:
Also, `knn` is not the name of the model. It is `algo`. But you can rename it above if you want.

I'm not sure if you can `vstack` a sparse matrix with another sparse matrix. We will see.